File: src/main/kotlin/easy/LongestCommonPrefix.kt
package easy

class LongestCommonPrefix {
    /*
     * Link: https://leetcode.com/problems/longest-common-prefix/
     * Difficult: Easy
     */

    fun longestCommonPrefix(strs: Array<String>): String {

        // 0 <= i <= 200
        var i = 0;
        var prefix = "";
        var sameLetter = true

        // Esse while vai de 0 até o tamanho da primeira palavra
        // Aumentar o tamanho da palavra não aumenta a quantidade, logo não é O(Nˆ2)
        // O(1)
        while(i < strs[0].length && sameLetter) {
            val letter = strs[0][i]

            // Esse for percorre todas as strings no array
            // O(N)
            for(str in strs) {
                if(i<str.length) {
                    sameLetter = str[i] == letter
                    if(!sameLetter) break
                    continue
                }
                sameLetter = false
                break
            }

            if(sameLetter){
                prefix += letter
            }

            i++
        }

        return prefix

    }
}

File: leetcode/easy/longest_common_prefix.cpp
#include "../data_structures.h"
#include <bits/stdc++.h>

using namespace std;

/*
 * Number: 14
 * Difficult: Easy
 * Link: https://leetcode.com/problems/longest-common-prefix/
 * Tags: String, Trie
 */ 

class LongestCommonPrefix {
public:
    /*
     * Time complexity: O(N+M)
     * Space complexity: O(M)
     * 
     * N is the size of array
     * M is the size of prefix
     * 
     * Time beats 5.78%
     * Memory beats 6.8%
     */
    string longestCommonPrefix(vector<string>& strs) {
        string aws = "";
        bool finish = false;

        for(int i = 0; i < strs[0].size(); i++) {
            for(string s: strs) {
                if(i >= s.size()) {
                    finish = true;
                    break;
                }

                if(strs[0][i] != s[i]) {
                    finish = true;
                    break;
                }      
            }

            if(finish) {
                break;
            }

            aws += strs[0][i];        
        }

        return aws;
    }
};